### DISCUSSION TITLE
Curious about the compilation process / working mechanism

### DISCUSSION THREAD

**Discussion by anzhi0708** [erg-lang/erg#416] [General]
Hello! 
**I am a beginner and very interested in programming and computer science.**
 I have always been curious about why there isn't another programming language that runs on the Python Virtual Machine; today, I accidentally discovered Erg. I understand that it can be compiled into Python bytecode and can also be translated into Python text code. Given that Python itself has modules like `compileall` that can directly compile Python code into bytecode, my question is: during the process of compiling Erg code into Python bytecode, is the code first translated into Python code and then compiled into bytecode? Or have you thoroughly studied Python bytecode, so that Erg code can be directly compiled into Python bytecode? (Considering that Erg has its own binary files, I lean towards the first guess.)

**Comment by anzhi0708** [comment]
I think I just found the answer [here](https://github.com/erg-lang/erg/blob/main/doc/EN/compiler/).

**Comment by mtshiba** [comment]
Erg generates Python bytecode from HIR (AST with type information etc.). The overall Erg architecture is described [here](https://github.com/erg-lang/erg/blob/main/doc/EN/compiler/architecture.md).

The backend that outputs Python scripts was added later and some features have not yet been implemented.

Here's my personal thought on why there aren't many transpiled languages that output Python bytecode.
Python bytecode is not publicly specified and compatibility breaks with each minor version. It is very difficult to follow this.
Currently Erg only supports Python 3.7~3.11.

**Reply by anzhi0708** [reply]
Thanks for your reply! ^_^
